### DISCUSSION TITLE
setNodes function not updating all nodes when used multiple times within Child component with ReactFlowProvider

### DISCUSSION THREAD

**Discussion by alikhani-dev** [xyflow/xyflow#3493] [Q&A] [answered]
Hi,
I'm facing an issue with using the `setNodes` function multiple times within a Child component while using `ReactFlowProvider`. It seems that only the last set of nodes is being updated (2 nodes in the example). However, when I use `setNodes` more than once in the App component, it works.

Here is an example showcasing the issues :
[codesandbox](https://codesandbox.io/s/sweet-tree-r9khm4?file=/Chid.js)

**Comment by bcakmakoglu** [comment, marked as answer]
What happens seems to be this:

The `setNodes` function of `useNodesState` under the hood uses `const [items, setItems] = useState(initialItems);`, so when you call `setNodes` it's just a `setState` function of `useState` - there's no steps in-between to change the state of the current nodes.

When you use `setNodes` of `useReactFlow` it's actually using something different, so you're not getting a `setState` function of a `useState` in this case, but instead a different function which will generate an array of *changes* which are then sent back to you via `onNodesChange` and then those changes are applied.
So I'm assuming the application of these changes happens *after* you do the second `setNodes` call in the `Child` component (in your sandbox example), thus changes sort of just overwrite each other (in reality, there's first a change-set sent back with the type `reset` and the updates you wanted to apply to the node and then right after another change-set is sent containing the next set of changes but the first ones haven't been applied yet, so the second one just "overwrites" the previous changes).

Not sure if this behavior can be remedied atm but to avoid this problem you could wrap subsequent `setNodes` calls in a timeout or sth. Not a nice solution and maybe there's a more stable way to achieve this.
